30413/Mihai Alexandru Sandu/Lab03/extra/main.c

```diff
@@ -0,0 +1,167 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+
+/*  *******************************************************
+
+    I wrote those functions and structures just to exercise
+
+    ******************************************************** */
+
+
+ typedef struct node
+    {
+        int value;
+        struct node *next;
+    }NodeT;
+
+typedef struct SLL
+{
+    int count;
+    NodeT *first;
+    NodeT *last;
+}SLListT;
+
+SLListT *createEmptySLList()
+{
+    SLListT *list = (SLListT*)malloc(sizeof(SLListT));
+    if (list)
+    {
+        list->count = 0;
+        list->first = list->last = NULL;
+    }
+    return list;
+}
+
+NodeT *createNode(int value)
+{
+    NodeT *p = (NodeT*)malloc(sizeof(NodeT));
+    if (p)
+    {
+        p->next = NULL;
+        p->value = value;
+    }
+    return p;
+}
+
+void AddFirst(SLListT *list, NodeT *p)
+{
+    if ( list )
+    {
+        if (list->first == NULL && list->last == NULL)
+        {
+            list->first = list->last = p;
+        }
+        else
+        {
+            p->next = list->first;
+            list->first = p;
+        }
+        list->count++;
+    }
+}
+
+// Function to find the intersection of two lists
+// input : 2 SLLists
+// output: the first node of the intersection
+
+NodeT *FindIntersection( SLListT *list1, SLListT *list2)
```

Comment:
It's a valid solution but not quite efficient. You've managed to do it in O(m*n) where m is the length of the first list and n is the length of the second.

There's a way to solve it in O(m+n).

Hint: use the length of the two lists then truncate the longer one.